### ISSUE TITLE
Installation problems in Ubuntu 16.04

### ISSUE THREAD

**Issue by rowyourboat** [jedisct1/dnscrypt-proxy#416]
Suggestions:
1. When DNScrypt-proxy is preparing to install, it also downloads the latest dnscrypt-resolvers.csv. 
2. Every time the service starts, it should test the resolvers until it finds one that works. If it selects at random even better.

Long version for inspiration:

I just did a fresh install of Ubuntu 16.04. I installed dnscrypt-proxy via Synaptic Package Manager, and the installation failed with errors:
"E: dnscrypt-proxy: subprocess installed post-installation script returned error exit status 1"
It tried to submit it's crash report, but all internet access failed. I had to uninstall the package to get Internet access back. I tried again, same result.

As a linux newbie, this concerned me. I really want DNScrypt, but it doesn't even work out of the box.
I tried again, this time looking at the install log.

The install details log suggests looking at:
systemctl status dnscrypt-proxy.service and journalctl -xe, so I did. It appears it's trying to use cisco as the server, and can't get the certificates, so the service shuts down, starts up, and tries again. Meanwhile Internet access on the machine is broken. Eventually I got it to work by pointing it to soltysiak.

Wonderful, I got it to work.  However, the point needs to be made, this makes the project look bad, and this is the kind of thing that makes people go back to Windows or Mac. Many people agree with the philosophy of Linux.  They like it a lot, but they can't use it. When they try to make the switch, they have to put their lives on hold to fuck around with all the unfinished buggy crap they run into as soon as they try to get started. Soon, they give up, and go back to Windows, as in my case.

Linux is very important to the computing world.  I love Linux, but I end up banging my head against the wall every time I try to switch. However, with the recent developments in the Windows OS, I'm determined to make it work this time.  I only learned of DNScrypt a few days ago, and now that I know about it, I consider it a critical piece of software for the future of secure computing.  This is something that everyone in the world should have running on their computer and phone.  If that's ever going to happen, it needs to JUST WORK as they say. It can be done.

My specific suggestion at this time is, when DNScrypt is preparing to install, gathering the dependencies, it also downloads the latest dnscrypt-resolvers.csv. That is basic and obvious, and it should be easy to implement.  Next, the program should test the resolvers until it finds one that works, before completing the install. You can have it select them at random if you like, until it finds one that works. In fact, it should do this check every time the service starts. By default. Simplicity for the users, guys.
This basic check will bring a lot more legitimacy to the project, and less headbanging will bring more people to Linux and keep them there. People with lives, and/or not in IT, are not going to open a terminal. And they should never have to.
Thanks for listening.  Sorry for the rant, but I'm on your side. I hope to see this implemented.

I also log fairly thoroughly my steps when I install an operating system.  I always end up reusing them later on when I want a fresh install, so that I can complete them very quickly without forgetting anything, and I can avoid problems I have run into in previous builds.  I volunteer my newbie-proof DNScrypt install notes for 16.04, once I have it up and running.  I would like to see it on the DNScrypt webpage.  Many websites I went to for instructions had me doing all kinds of crazy stuff in the command line, but I just eventually found in sitting there in the Synaptic package manager. I would like to make sure anyone coming in behind me finds it right where they need it and have it up and running in a few minutes (like you can in Windows), versus fucking around all day with out-of-date methods.

**Comment by licaon-kter** [comment]
/close as invalid

TL;DR, wrong place to rant, report this to the packagers where it belongs: https://bugs.launchpad.net/ubuntu/+source/dnscrypt-proxy/+bug/1516420

**Comment by rowyourboat** [comment]
OK, I will resubmit without rant.  However, this isn't just about the package, this is about the core functionality of dnscrypt-proxy, isn't it? I would have the same problem if installed through another method other than synaptic, and if I had done it on Fedora or something else.

**Comment by licaon-kter** [comment]
Again, if you rely on a distro package, go rant there, the problem is not here.

Actually this is a multipart problem, as servers go up/down pretty random, so a package maintainer would need to repackage it pretty often, not doable, and this needs a script to update and check the signature of the .cvs.

Regarding your issue, edit the .cvs file and remove the :443 port address at cisco entries, although better grab the latest version and replace it.

**Comment by jedisct1** [comment]
@licaon-kter don't be so rude, please.

dnscrypt-proxy was not specially written for Ubuntu. In fact, I don't even use Ubuntu at all and I didn't even know that they provide a package for it.

It's a generic application, that was designed to work on many different operating systems. It's just an application. It doesn't include any installer, package or distribution-specific scripts.

There are so many operating systems and distributions that it would not make sense. Instead, it provides the application, and distribution-specific packages are made by packages maintainers who know best of to integrate new applications into specific operating systems or distributions.

You are absolutely correct. Installers/packages should retrieve the latest version of the CSV file, verify it, and shouldn't hard-code a resolver that may not match your expectations, or that may not even work.

The Ubuntu package is not maintained by the dnscrypt maintainers, but by Ubuntu maintainers. And the issues you describe are really about the package itself.

The good news is that Ubuntu maintainers are usually reactive and helpful. The bad news is that they don't use rolling releases, so you'll have to wait until the next release before bugs like those you describe get fixed.

No matter what, you should indeed report these issues to Ubuntu forums, or to the Ubuntu bug tracker.

**Comment by licaon-kter** [comment]
Sorry to come of as rude, the wonders of internet communication. :) <- there I fixed it

Just trying to be succinct and answer to the issue at hand, although I did not find that funny the whole wall-of-text asking for "Ubuntu or bust" ;)

**Comment by rowyourboat** [comment]
I'm trying to say that even if I download it from DNScrypt.org it should do these things, regardless of the OS, Linux, Android, Window, Mac.
That's why it should be build directly into the program, right here on github.

@licaon-kter Not sure where you get the ubuntu or bust idea. I certainly don't feel that way.

**Comment by licaon-kter** [comment]
I read it like this: "I've installed it in Ubuntu and it did not work... here are the steps to fix it", since instead of just looking at the package page -> Bug reports -> and see *exactly* one issue, yours, you've started to write an 3902 characters long post about Windows and what not. :-|

You've made the choice of the distro (services running like dnsmasq), the install method (outdated resolvers list) and so forth.

**Comment by jedisct1** [comment]
Changing the network settings, especially DNS settings, is hard to do in a portable way. Even just on Linux. Some distros use systemd, some don't, some run dnsmasq by default, some run unbound, some have custom scripts to update `/etc/resolv.conf`.

So I wouldn't expect an app like dnscrypt to mess with network settings or with the configuration of other apps. If only because this can realistically only be tested on a very small subset of existing OS/distributions/versions. So, it will break.

And yes, that's one of the reasons why OSX and Windows remains vastly superior as desktop operating systems. But we're digressing here.

**Comment by rowyourboat** [comment]
@licaon-kter You are the kind of "linux guy" that will never get it.  Please go away.
@jedisct1 Thank you for your quick responses and remaining constructive. Is it not possible to have it download the latest resolvers? It is not possible to do a fetch certificate check at service startup?
Perhaps I don't understand the delineation of what is here at github and what is added by packagers.
What exactly does the software package here at github do without customizations provided by OS maintainers?

**Comment by jedisct1** [comment]
The source code include a script to update the list of resolvers: https://github.com/jedisct1/dnscrypt-proxy/blob/master/dnscrypt-update-resolvers.sh.in

What you have on Github is for developers and package maintainers.

**Comment by licaon-kter** [comment]
@rowyourboat Sorry to hear that, here, have a link: http://linux.oneandoneis2.org/LNW.htm
Hope that would clear you misreading of my responses.

**Comment by rowyourboat** [comment]
OK, let's focus on having the service verify that it can fetch the certificate before it starts, and if it can't try another resolver? Is that something that could be handled by the code here?
